File: golang/1091. 二进制矩阵中的最短路径.go
func shortestPathBinaryMatrix(grid [][]int) int {
	if grid[0][0] == 1 {
		return -1
	}
	n := len(grid)
	dist := make([][]int, n)
	for i := 0; i < n; i++ {
		dist[i] = make([]int, n)
		for j := 0; j < n; j++ {
			dist[i][j] = 0x3f3f3f3f
		}
	}
	q := [][2]int{{0, 0}}
	dist[0][0] = 1
	for len(q) > 0 {
		x, y := q[0][0], q[0][1]
		q = q[1:]
		if x == n-1 && y == n-1 {
			return dist[x][y]
		}
		for dx := -1; dx <= 1; dx++ {
			for dy := -1; dy <= 1; dy++ {
				if x+dx < 0 || x+dx >= n || y+dy < 0 || y+dy >= n { // 越界
					continue
				}
				if grid[x+dx][y+dy] == 1 || dist[x][y]+1 >= dist[x+dx][y+dy] { // 单元格值不为 0 或已被访问
					continue
				}
				dist[x+dx][y+dy] = dist[x][y] + 1
				q = append(q, [2]int{x + dx, y + dy})
			}
		}
	}
	return -1
}

File: 刷题/cpp/1091. 二进制矩阵中的最短路径.cpp
class Solution {
public:
    struct Point{
        int x;
        int y;
        int step;
    };
    int dir[8][2]={{-1,0},{-1,1},{0,1},{1,1},{1,0},{1,-1},{0,-1},{-1,-1}};
    bool check(int x,int y,int m,int n){
        if(x>=0&&x<m&&y>=0&&y<n){
            return true;
        }
        return false;
    }
    int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
        int m=grid.size(),n=grid[0].size();
        queue<Point> qu;
        if(grid[0][0]==1){
            return -1;
        }
        qu.push(Point{0,0,0});
        vector<vector<int> > inq(m,vector<int>(n));
        inq[0][0]=1;
        while(!qu.empty()){
            Point now=qu.front();
            qu.pop();
            if(now.x==m-1&&now.y==n-1){
                return now.step+1;
            }
            for(int i=0;i<8;i++){
                int tx=dir[i][0]+now.x;
                int ty=dir[i][1]+now.y;
                if(check(tx,ty,m,n)&&inq[tx][ty]==0&&grid[tx][ty]==0){
                    qu.push(Point{tx,ty,now.step+1});
                    inq[tx][ty]=1;
                }
            }
        }
        return -1;
    }
};